File: Algorithm/Recursion/118.py
# Pascal's Triangle


# Recursion
class Solution:
    def generate(self, numRows):
        if numRows == 1:
            return [[1]]

        triangle = self.generate(numRows - 1)

        prev_row = triangle[-1]
        cur_row = [1]

        for i in range(1, len(prev_row)):
            cur_row.append(prev_row[i - 1] + prev_row[i])
        cur_row.append(1)

        triangle.append(cur_row)
        return triangle

File: Algorithm/Recursion/118.ts
function generate2(numRows: number): number[][] {
  if (numRows === 1) {
    return [[1]];
  }

  const triangle: number[][] = generate2(numRows - 1);
  const prevRow: number[] = triangle[triangle.length - 1];
  const curRow: number[] = [1];

  for (let i: number = 1; i < prevRow.length; i++) {
    curRow.push(prevRow[i - 1] + prevRow[i]);
  }
  curRow.push(1);
  triangle.push(curRow);
  return triangle;
}
